konduit-serving-vertx-protocols/konduit-serving-http/src/main/java/ai/konduit/serving/vertx/protocols/http/api/InferenceHttpApi.java

```diff
@@ -0,0 +1,95 @@
+ *  * under the License.
+ *  *
+ *  * SPDX-License-Identifier: Apache-2.0
+ *  *****************************************************************************
+ */
+
+package ai.konduit.serving.vertx.protocols.http.api;
+
+import ai.konduit.serving.pipeline.api.data.Data;
+import ai.konduit.serving.pipeline.api.pipeline.PipelineExecutor;
+import ai.konduit.serving.pipeline.util.ObjectMappers;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.ext.web.RoutingContext;
+import lombok.AllArgsConstructor;
+import org.nd4j.shade.guava.base.Strings;
+
+import java.nio.charset.StandardCharsets;
+
+import static io.netty.handler.codec.http.HttpHeaderValues.APPLICATION_JSON;
+import static io.netty.handler.codec.http.HttpHeaderValues.APPLICATION_OCTET_STREAM;
+import static io.vertx.core.http.HttpHeaders.ACCEPT;
+import static io.vertx.core.http.HttpHeaders.CONTENT_TYPE;
+
+@AllArgsConstructor
+public class InferenceHttpApi {
+
+    PipelineExecutor pipelineExecutor;
+
+    private Data extractData(String contentType, RoutingContext ctx) {
+        try {
+            if (contentType.contains(APPLICATION_JSON.toString())) {
+                return Data.fromJson(ctx.getBodyAsString());
+            } else if (contentType.contains(APPLICATION_OCTET_STREAM.toString())) {
+                return Data.fromBytes(ctx.getBody().getBytes());
+            } else {
+                throw new KonduitServingHttpException(HttpApiErrorCode.INVALID_CONTENT_TYPE_HEADER,
+                        String.format("Invalid Content-Type header %s. Should be one of [application/json, application/octet-stream]", contentType));
+            }
+        } catch (Exception exception) {
+            throw new KonduitServingHttpException(HttpApiErrorCode.DATA_PARSING_ERROR, exception.toString());
+        }
+    }
+
+    //TODO: add swagger related annotations to this method or update this class for better swagger annotations support
+    public void predict(RoutingContext ctx) {
+        String contentType = ctx.request().headers().get(CONTENT_TYPE);
+        String accept = ctx.request().headers().get(ACCEPT);
+
+        if(Strings.isNullOrEmpty(contentType)) {
+            throw new KonduitServingHttpException(HttpApiErrorCode.MISSING_OR_EMPTY_CONTENT_TYPE_HEADER,
+                    "Content-Type header should not be null. Possible values are: [application/json, application/octet-stream]");
+        }
+
+        if(Strings.isNullOrEmpty(accept)) {
+            throw new KonduitServingHttpException(HttpApiErrorCode.MISSING_OR_EMPTY_ACCEPT_HEADER,
+                    "Accept header should not be null. Possible values are: [application/json, application/octet-stream]");
+        }
+
+        Data input = extractData(contentType, ctx);
+        Data output;
+
+        try {
+            output = pipelineExecutor.exec(input);
+        } catch (Exception exception) {
+            throw new KonduitServingHttpException(HttpApiErrorCode.PIPELINE_PROCESSING_ERROR, exception.toString());
```

Comment:
This hides part of the exception stack trace.
Here's InferenceVerticleHttpTest (after I've added a failure):
```
java.lang.IllegalStateException: Intentionally causing a pipeline execution failure!
	at ai.konduit.serving.vertx.protocols.http.test.FailureTestingPipelineStep$Runner.exec(FailureTestingPipelineStep.java:69)
	at ai.konduit.serving.pipeline.impl.pipeline.SequencePipelineExecutor.exec(SequencePipelineExecutor.java:69)
	at ai.konduit.serving.vertx.protocols.http.api.InferenceHttpApi.predict(InferenceHttpApi.java:75)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1034)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:131)
```

And here's what's actually logged:
```
23:14:09.422 [vert.x-eventloop-thread-0] ERROR a.k.s.v.p.h.v.InferenceVerticleHttp - Request failed with cause 
ai.konduit.serving.vertx.protocols.http.api.KonduitServingHttpException: Error Code: PIPELINE_PROCESSING_ERROR
 Error Message: java.lang.IllegalStateException: Intentionally causing a pipeline execution failure!
	at ai.konduit.serving.vertx.protocols.http.api.InferenceHttpApi.predict(InferenceHttpApi.java:78)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1034)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:131)
```

Note the loss of the info about the actual pipeline step (and method) that it occurred in:
`at ai.konduit.serving.vertx.protocols.http.test.FailureTestingPipelineStep$Runner.exec(FailureTestingPipelineStep.java:69)`

Reply:
Great catch! Thanks!